Proposal: Allow coercing ref to anonymous array init to `[*]` and `[*:x]`

```zig
const std = @import("std");
const expectEqual = std.testing.expectEqual;

test "ref to anon array init coerced to slice" {
    const values: []const i32 = &.{ 1, -2 };
    try expectEqual(1, values[0]);
    try expectEqual(-2, values[1]);
    try expectEqual(2, values.len);
}

test "ref to anon array init coerced to sentinel slice" {
    const values: [:0]const i32 = &.{ 3, -4 };
    try expectEqual(3, values[0]);
    try expectEqual(-4, values[1]);
    try expectEqual(0, values[2]);
    try expectEqual(2, values.len);
}

test "ref to anon array init coerced to many-ptr" {
    const values: [*]const i32 = &.{ 5, -6 };
    try expectEqual(5, values[0]);
    try expectEqual(-6, values[1]);
}

test "ref to anon array init coerced to sentinel many-ptr" {
    const values: [*:0]const i32 = &.{ 7, -8 };
    try expectEqual(7, values[0]);
    try expectEqual(-8, values[1]);
    try expectEqual(0, values[2]);
}
```

The latter two tests currently do not compile:

```
repro.zig:20:36: error: type 'i32' does not support array initialization syntax
    const values: [*]const i32 = &.{ 5, -6 };
                                  ~^~~~~~~~~
repro.zig:26:38: error: type 'i32' does not support array initialization syntax
    const values: [*:0]const i32 = &.{ 7, -8 };
                                    ~^~~~~~~~~
```

Coercions from ref-to-anon-array-inits to slices are already supported, so I don't see why coercions to many-pointers shouldn't be. There is no ambiguity or risk of error at play and adding support for this is as simple as extending a single check in Sema.

Example of actual code of mine interacting with a C API where I'm currently running into this limitation and need to introduce awkward `@as` coercions:

```zig
fn glShaderSource(shader: c_uint, count: c_int, string: [*]const [*]const u8, length: ?[*]const c_int) void {}
fn glClearBufferfv(buffer: c_uint, drawbuffer: c_int, value: [*]const f32) void {}

const shader_source_preamble = "#version 410 core\n";
const vertex_shader_source = "void main() { gl_Position = vec4(0, 0, 0, 1); }\n";

glShaderSource(
    vertex_shader,
    2,
    &@as([2][*]const u8, .{ shader_source_preamble, vertex_shader_source }),
    &@as([2]c_int, .{ @intCast(shader_source_preamble.len), @intCast(vertex_shader_source.len) }),
);

glClearBufferfv(GL_COLOR, 0, @as([]const f32, &.{ 1, 1, 1, 1 }).ptr);
```

Supporting coercions from ref-to-anon-array-inits to many-pointers would make code like this much nicer to both read and write:

```diff
 glShaderSource(
     vertex_shader,
     2,
-    &@as([2][*]const u8, .{ shader_source_preamble, vertex_shader_source }),
-    &@as([2]c_int, .{ @intCast(shader_source_preamble.len), @intCast(vertex_shader_source.len) }),
+    &.{ shader_source_preamble, vertex_shader_source },
+    &.{ @intCast(shader_source_preamble.len), @intCast(vertex_shader_source.len) },
 );

-glClearBufferfv(GL_COLOR, 0, @as([]const f32, &.{ 1, 1, 1, 1 }).ptr);
+glClearBufferfv(GL_COLOR, 0, &.{ 1, 1, 1, 1 });
```

--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -4557,14 +4557,20 @@ fn zirValidateArrayInitRefTy(
     };
     const ptr_ty = maybe_wrapped_ptr_ty.optEuBaseType(mod);
     assert(ptr_ty.zigTypeTag(mod) == .Pointer); // validated by a previous instruction
-    if (ptr_ty.isSlice(mod)) {
-        // Use array of correct length
-        const arr_ty = try mod.arrayType(.{
-            .len = extra.elem_count,
-            .child = ptr_ty.childType(mod).toIntern(),
-            .sentinel = if (ptr_ty.sentinel(mod)) |s| s.toIntern() else .none,
-        });
-        return Air.internedToRef(arr_ty.toIntern());
+    switch (mod.intern_pool.indexToKey(ptr_ty.toIntern())) {
+        .ptr_type => |ptr_type| switch (ptr_type.flags.size) {
+            .Slice, .Many => {
+                // Use array of correct length
+                const arr_ty = try mod.arrayType(.{
+                    .len = extra.elem_count,
+                    .child = ptr_ty.childType(mod).toIntern(),
+                    .sentinel = if (ptr_ty.sentinel(mod)) |s| s.toIntern() else .none,
+                });
+                return Air.internedToRef(arr_ty.toIntern());
+            },
+            else => {},
+        },
+        else => {},
     }
     // Otherwise, we just want the pointer child type
     const ret_ty = ptr_ty.childType(mod);

--- a/test/behavior/array.zig
+++ b/test/behavior/array.zig
@@ -804,6 +804,52 @@ test "slice initialized through reference to anonymous array init provides resul
     try std.testing.expectEqualSlices(u16, &.{ 123, 456, 123, 456 }, foo);
 }
 
+test "sentinel-terminated slice initialized through reference to anonymous array init provides result types" {
+    var my_u32: u32 = 123;
+    var my_u64: u64 = 456;
+    _ = .{ &my_u32, &my_u64 };
+    const foo: [:999]const u16 = &.{
+        @intCast(my_u32),
+        @intCast(my_u64),
+        @truncate(my_u32),
+        @truncate(my_u64),
+    };
+    try std.testing.expectEqualSentinel(u16, 999, &.{ 123, 456, 123, 456 }, foo);
+}
+
+test "many-item pointer initialized through reference to anonymous array init provides result types" {
+    var my_u32: u32 = 123;
+    var my_u64: u64 = 456;
+    _ = .{ &my_u32, &my_u64 };
+    const foo: [*]const u16 = &.{
+        @intCast(my_u32),
+        @intCast(my_u64),
+        @truncate(my_u32),
+        @truncate(my_u64),
+    };
+    try expectEqual(123, foo[0]);
+    try expectEqual(456, foo[1]);
+    try expectEqual(123, foo[2]);
+    try expectEqual(456, foo[3]);
+}
+
+test "many-item sentinel-terminated pointer initialized through reference to anonymous array init provides result types" {
+    var my_u32: u32 = 123;
+    var my_u64: u64 = 456;
+    _ = .{ &my_u32, &my_u64 };
+    const foo: [*:999]const u16 = &.{
+        @intCast(my_u32),
+        @intCast(my_u64),
+        @truncate(my_u32),
+        @truncate(my_u64),
+    };
+    try expectEqual(123, foo[0]);
+    try expectEqual(456, foo[1]);
+    try expectEqual(123, foo[2]);
+    try expectEqual(456, foo[3]);
+    try expectEqual(999, foo[4]);
+}
+
 test "pointer to array initialized through reference to anonymous array init provides result types" {
     var my_u32: u32 = 123;
     var my_u64: u64 = 456;
@@ -815,6 +861,19 @@ test "pointer to array initialized through reference to anonymous array init pro
         @truncate(my_u64),
     };
     try std.testing.expectEqualSlices(u16, &.{ 123, 456, 123, 456 }, foo);
+}
+
+test "pointer to sentinel-terminated array initialized through reference to anonymous array init provides result types" {
+    var my_u32: u32 = 123;
+    var my_u64: u64 = 456;
+    _ = .{ &my_u32, &my_u64 };
+    const foo: *const [4:999]u16 = &.{
+        @intCast(my_u32),
+        @intCast(my_u64),
+        @truncate(my_u32),
+        @truncate(my_u64),
+    };
+    try std.testing.expectEqualSentinel(u16, 999, &.{ 123, 456, 123, 456 }, foo);
 }
 
 test "tuple initialized through reference to anonymous array init provides result types" {

